### DISCUSSION TITLE
Simulated orbit gets async with input time series for longer periods (e.g. 31 days)

### DISCUSSION THREAD

**Discussion by zinphi** [groops-devs/groops#211] [Q&A] [answered]
### Description

When using the program SimulateOrbit for simulating an orbit over a longer period (e.g., 31 days), I can see an increasing time lag between the epochs of the integrated orbit and the input time series (in my case a simple uniform time series with 5s sampling). The last epoch in the orbit file (after 31 days) has the value (MJD) 52306.000000000262559752 which is about 2.3µs off with respect to the last epoch in the uniform time series (=52306). 2.3µs correspond already to an 18cm displacement for a LEO satellite. Numerically, I cannot really comprehend this, because the integrator should anyway interpolate the result on the requested epochs when using dynamic time steps. Hence, to my understanding, the result should refer exactly (up to double precision accuracy) to the input epochs. 

### GROOPS version

main (latest commit)

### Operating systems

- [X] Linux
- [ ] Windows
- [ ] macOS
- [ ] Other (please specify in the description)

**Comment by tmayerguerr** [comment]
Hi Philipp,

I'm not sure what level of calculation precision you expect. The time difference between (MJD) 52306.000000000262559752 and 52306 is already on the border of double precision (15 digits). GROOPS internally uses a time format where the integer part is stored extra so that the full double precision is available for the fractional part.

To explain the difference in the integrator: 
The orbit integrator uses the start time and the sampling from the time series. The next time step is determined by adding the sampling. This means that after half a million steps, the error can add up to a few microseconds. Unfortunately, this behavior cannot simply be changed. I also do not think that this would increase the accuracy of the integrated positions.

You can interpolate to the desired times with [InstrumentResample](https://groops-devs.github.io/groops/html/InstrumentResample.html).

regards
Torsten

**Reply by zinphi** [reply]
Hello Torsten,

Many thanks for your swift answer. Ok, I can understand now the nature of the problem. However, it doesn't explain to me (1) why a relative timescale is used (at least for the final interpolation step) and (2) why a relative time scale should lead to a deterioration of the integrated time scale in general. 
To explain my second point: if you have a small deltaT which you can precisely calculated (e.g., in case of a uniform sampling), let's say 1e-3s, then you can assume a numerical (timing) error per integration step of about 1e-19s when using double precision to represent your relative time. When you do the integration now let's say 2.6784e+09 (=86400*31/1e-3) times, your worst case rounding error is about 3e-10s which is still in the sub ns range. 3e-10s, on the other hand, is about the numerical error which you would get when you use an absolute time scale from the beginning of the integration (which has after 31 days already a magnitude of about 3e6s). Thus, you should not loose timing precision just by using a relative time scale. Just if the calculation of this relative time scale is inaccurate (e.g., when derived through a single subtraction) you will loose precision. 
But even if you derive your deltaT by deltaT(i)=t(i+1)-t(i) (which you have to, since you support non uniform timeseries), you will not come up with an integration error since your individual deltaT(i) would 'preserve' the cancellation error you introduce through the subtraction. You can easily verify this by running this simple Matlab code which builds deltaTs from an absolute scale and sums them up again:
```
dt0=1/70;
t0=0:dt0:(86400*31);
%cancellation is introduced in this step
dt=diff(t0);
%cancellation 'cancels' out again
t1=cumsum(dt);
%no rounding error is introduced when you sum them up
roundError=t1(end)-t0(end); %=0s
%here is what happens if you use a single value for summation (with limited precision)
roundErrorSingle=dt(end)*(numel(t1))-t0(end); %=0.015s
%here is what happens if you use a single value for summation (with high precision)
roundErrorSinglePrecise=dt0*(numel(t1))-t0(end); %=0s
```

So, I'm still not quite sure what actually happens in the integrator and if it is correct to simply perform an interpolation afterwards. I personally would consider a subsequent interpolation a somewhat dirty solution to this since it will add additional numerical errors on top (possibly even in the higher frequencies which are crucial for gravity field recovery and the additional error might even not be stationary since the distance to the nearest support point increases with time).

Best regards,
Philipp

**Comment by tmayerguerr** [comment]
Hi Philipp,

First of all: SimulateOrbit supports only time series with constant sampling. If an irregular time series is provided an error message is shown. Ok, here we should improve the documentation.

The integrator API expects timeStart and sampling as arguments, which might no be the best choice but cannot easily be changed. The sampling is determined by (following your matlab example)
  ```  dt = median(diff(t))```
Therefore it is something between your two extreme cases.

If you want avoid resampling, you might restart SimulateOrbit every day (With daily [LoopPgramm](https://groops-devs.github.io/groops/html/LoopPrograms.html) around and the last orbit as new integrationConstants->file)

regards
Torsten

**Comment by zinphi** [comment]
Dear Torsten,

Thanks again for your explanation. This makes things again clearer to me. Just as an idea, would it be possible to change the way to determine the sampling to:
`dt = (t(end) - t(1))/(numel(t)-1)`
In this way one would again avoid to calculate a difference of two closeby values which leads in the end to a maximized cancellation error. With this modified formulation your final integrated timeseries should be identical again with the original nearly up to working precision.

Best regards,
Philipp

**Comment by tmayerguerr** [comment]
Dear Philipp,

Yeah, maybe this helps. Please try to replace line 125 in source/programs/simulation/simulateOrbit.ccp by
```Time sampling  = 1./(times.size()-1) * (times.back()-times.front());```

Torsten

**Comment by tmayerguerr** [comment, marked as answer]
A new version is in main, which should improve the numerical precision.
Torsten

**Reply by zinphi** [reply]
Dear Torsten,
Thanks for your engagement and for quickly resolving this issue! Yesterday, I briefly tested your suggested solution (replacing the single line) myself but I did still get the same deviations. However, your more elaborate fix did the trick and I now get time series which are really close to working precision. Maybe as a last suggestion for somewhen in the future: The input/output time series are now really close together but still not identical. Since, on this scale, these deviations do not matter, I would propose to simply replace the output time series with the input time series for maximum consistency.

Best,
Philipp
